### DISCUSSION TITLE
microphone preset for EasyEffects

### DISCUSSION THREAD

**Discussion by noelcurray** [wwmm/easyeffects#1227] [Q&A] [answered]
Any website recommendation on where I can get microphone preset for EasyEffects?

**Comment by wwmm** [comment]
Although I apply effects to my microphone frequently I never searched for presets on the web. This kind of thing is usually too personal for presets to be useful. What kind of thing are you trying to do?

**Reply by noelcurray** [reply]
I want to make my microphone to sound better. I bought a condenser microphone and a phantom power for it but it still sound like an ordinary mic but just louder. I want to make it sound like a radio dj. I don't know if its possible with just easyeffects and without editing software.

**Reply by polarathene** [reply]
I am also trying to learn how to configure EasyEffects for better control over the mic output.

It would be good to know what are common effects to use, and their order or advice on tuning the settings (_where defaults may not be as appropriate for mic input?_).

## Appropriate effects and advice?

### Gate and Noise Reduction

I have currently gotten familiar with the "Gate" to better handle where "Noise Reduction" doesn't do well, they seem complimentary at controlling what/when input goes through to be handled, and I have read that "Noise Reduction" effect is simpler than NoiseTorch equivalent (_despite both using RNNoise underneath_), so perhaps it is meant to be paired with "Gate", but I am not sure if it makes more sense for the Gate to come before or after, it seems to be better as when positioned afterwards.

### Echo Canceller

I'm using headphones, so "Echo Canceller" might not make a difference for me (_I do hear a bit of delay/echo, but I think that's partially due to bluetooth headphones, or it's active-noise-cancellation feature that can't be disabled, the microphone is a separate device over USB_). I would still be curious where it's appropriate to place that effect.

### Equalizer

I still need to get practice with the "Equalizer", which is a bit harder to follow guides online that I come across but I'll figure that out, I assume it should be positioned somewhere after the "Gate".

### Autogain

I then have "Autogain" which I've been experimenting with. It seems to handle normalizing my voice level without me having to crank up my microphones gain knob to max and speak right up close (dynamic mic), which would also pick up less desired sounds... however sometimes the "Autogain" will do a big spike on the _Output Gain_ value, boosting any quiet noise (_that seeps through once "Noise Reduction" / "Gate" are active_) this is when no speaking or other loud noise is present (_at least with the "Gate" / NR effects off for testing_). The sound of breathing or taking a drink can be heard despite not being near the mic and it's lower gain level knob setting. That is until a loud enough sound comes into play and reduces the _Output Gain_ momentarily.

Not sure if this is a bug with "Autogain" as _Output Gain_ in that situation can otherwise stay high and doesn't drop which is probably to be expected. While experimenting with the various tunables of the effect however, this made it difficult, as "Reset" would not reset the _Output Gain_, only a loud enough noise would.

Is `-23db` appropriate for mic input as a "Autogain" target value? My understanding is that will adjust the input volume by _Output Gain_ to reach that volume level, in the quiet background noise scenario this was 30-40db+ _Output Gain_, sometimes larger, which was unpleasant. The default settings for that scenario also did not contrast well to a sudden loud noise like a clap unless the history time was raised until around 30s.

### End of list/chain - Autogain + Limiter?

I assume "Autogain" belongs near the end of the list. I've read that a "Limiter" is worthwhile to add as well (_my mic has one built in when the gain is over 50%, but at least with "Autogain" in play I assume it makes sense to include a "Limiter" effect at the end as well?_).

There's an _Auto Leveling_ option in "Limiter" effect, but it just seems to add more options (_at least I assume the others are somehow still applied_), I've yet to try this effect out or look into it yet. I am a little worried about messing up a setting while experimenting that would harm my ears via the headphones though, so figuring out how to setup the Limiter for that is my next go-to.

Some of the Autogain experiments did boost the audio levels excessively, and when I mixed in the Equalizer after it, some really weird sounds.

### Misc (Side-chain, low/high pass filters)

Some effects have side-chain settings and low/high pass filters that I can read up about elsewhere. I'm not sure how appropriate that is for tuning mic input for voice calls rather than music vocals.

For now I use the EasyEffects _Internal_ input, rather than the _External_ input source of my mic. Not sure when / if that'd be suitable to change.

Still plenty of other effects to explore and figure out if they're useful for mics.

---

## Resources / Guides?

That's where I am presently, lot to learn and figure out, some specific to EasyEffects and how/what it offers for configuring effects. The in-app manual is helpful (Early on I tried just changing values)

**Presets** would at least provide a rough guide overview. I'll continue reading and looking for resources though :+1:

**Reply by wwmm** [reply]
>and I have read that "Noise Reduction" effect is simpler than NoiseTorch equivalent (despite both using RNNoise underneath), so perhaps it is meant to be paired with "Gate",

I think NoiseTorch has additional processing done on top of RNNoise. On EasyEffects some people combine the gate and the RNNoise to have some kind of "activity detection" besides the noise reduction. Depending othe the signal level and type of noises there may be some difference depending on whether the gate is before or after rnnoise but I would not expect the difference to be big.

>I'm using headphones, so "Echo Canceller" might not make a difference for m

I have a headphone where its microphone can capture some of the audio played by its speakers. Probably a design flaw. So there are cases where it makes sense to have a echo canceller when using headphones. That being said our current echo canceller isn't as good as I would like.

I think it is probably better to put the echo canceller at the beginning of the pipeline when filtering the microphone signal so the filter is as close to the microphone as possible.

>sometimes the "Autogain" will do a big spike on the Output Gain value, boosting any quiet noise

I usually use it only in the speakers pipeline. So I am not sure about how it is behaving when processing microphones. But it may be a good idea to put a gate before it to cut the mic signal once there is no voice activity. 

>Not sure if this is a bug with "Autogain" as Output Gain in that situation can otherwise stay high and doesn't drop which is probably to be expected. While experimenting with the various tunables of the effect however, this made it difficult, as "Reset" would not reset the Output Gain, only a loud enough noise would.

It depends on the mode and also on the `maximum history`. The reset should reset all loudness estimations. What in turn should have had some kind of effects on the output gain. Strange...

>I assume it makes sense to include a "Limiter" effect at the end as well?

Yes. I usually have a limiter as the last filter and the autogain before it. This way the limiter can protect the soundcard if the autogain does something odd. Limiting signals properly isn't something easy to do. So reinventing the wheel by replicating in the autogain what limiters do internally is not very viable. It is easier to put the proper tool after the autogain.

>For now I use the EasyEffects Internal input, rather than the External input source of my mic. Not sure when / if that'd be suitable to change.

The external sidechain input is useful in some corner cases. You probably have seen that when some streamers talk the volume of background sound in their computers is automatically reduced. You can achieve the same thing if you put in your speakers pipeline a compressor with the sidechain input set to your microphone. This way when you speak the volume of whatever you are listening to will be automatically reduced.

Musicians and audio engineers obviously have more advanced use cases for this. But for ordinary people this is probably the most common use case.

**Reply by polarathene** [reply]
Most of this feedback was on the same day as your reply.

I sinced played with the settings a bit more, and have been told the audio quality was good on the listener end of a group meeting in Discord (_disabled the similar discord app audio features_). I'll share those in a follow-up reply for other readers landing here :)

### RNNoise + Gate

> Depending othe the signal level and type of noises there may be some difference depending on whether the gate is before or after rnnoise but I would not expect the difference to be big.

I've noticed it's better to place RNNoise prior to the gate. Otherwise since it is trained for "listening" to the environment to react, it doesn't seem to handle responding well to a gate suddenly opening with the transition from silence.

It's also easier to see with the gate out of the way, that while it appears to block out sounds, some leakage still occurs perhaps due to noise dampening instead? Which gets countered by effects that raise the gain such as "Autogain".

When the Gate comes afterwards, this amplification is avoided (depending on the input). I've had it sometimes open the gate where RNNoise would suppress it, so having it before the Gate better controlled when it should actually open.

**UPDATE:** For those curious, there is some unknown (_probably documented somewhere_) duration that RNNoise adapts to the environment. Eg a small container with pills can be shaked and that sudden spike in volume/frequency gets let through opening a Gate, but if the noise continues for a while, RNNoise temporarily realizes it's environmental and starts to ignore it until it forgets after some time without the noise present. That does make it a little tricky to test, especially with the Gate.

When RNNoise is allowing sound through, it may dampen it to some extent, but some audio like music can partially seep through sounding broken/weird, as although your Gate might be open, RNNoise cannot be toggled off (_which is fine, it will still suppress certain noises well, but other sounds like clicking/tapping of keys will be more audible until RNNoise "closes"_).

### Echo Canceller

> I have a headphone where its microphone can capture some of the audio played by its speakers. Probably a design flaw.

> I think it is probably better to put the echo canceller at the beginning of the pipeline when filtering the microphone signal so the filter is as close to the microphone as possible.

Makes sense, thanks :+1: 

### Autogain

> I usually use it only in the speakers pipeline. So I am not sure about how it is behaving when processing microphones. But it may be a good idea to put a gate before it to cut the mic signal once there is no voice activity.

I wasn't sure about if the "Compressor" or "Autogain" was suitable for having before a Gate. The "Compressor" was neat since it allowed to have more flexible distance from the dynamic mic, but both can result in quiet sounds being normalized/boosted too high to trigger the gate anyway :sweat_smile:

I suppose it'd be useful to apply some sort of falloff/intensity curve based on current dB - which is what a Gate seems to handle, but presently limited to a single Gate effect.

The main issue with "Autogain" was as mentioned, with _Output Gain_ being difficult to lower/reset.

My Gate could close and it'd remain up for quite some time beyond the history (which isn't surprising since neither of the "reset" buttons for that effect changed the _Output Gain_ value).

> It depends on the mode and also on the maximum history. The reset should reset all loudness estimations.
> What in turn should have had some kind of effects on the output gain. Strange...

Yeah, I've tried changing the modes and history (eg: setting it back to 6s and waiting), but I didn't notice it making a difference. Perhaps it was adjusting for the audio, and the UI itself for _Output Gain_ wasn't updating?

<details>
<summary>Original message</summary>

I'm not sure how to reproduce that as sometimes it would fluctuate fine (when the bar was resizing), but it seems to have more trouble when over 30 and overdrawing beyond the bars actual bounds. I remember giving up on it last night and when I came back to it much later, it had at some point gone back down to a small value.

It's a bit hard to test that as loud sounds like clapping bother those around me, but I found it would always go back to a changing value if a loud enough sound caused "Autogain" to adjust for the... ceiling/peak being higher? Otherwise it's stuck keeping _Output Gain_ boosted, won't adjust to a lower value over long duration of silence without that which I don't think is intentional.

I had tried setting a target level of `-60db` (instead of the default `-23db`) when this bug happened, thinking that if it's trying to raise the quietness (without Gate) to `-23db`, and that's why it won't lower _Output Gain_.

</details>

<details>
<summary>Some screenshots and running Autogain for a while</summary>

Here's a screenshot I just took with the gate blocking out all noise, been over a minute and _Output Gain_ hasn't changed:
![Screenshot_20220830_170038](https://user-images.githubusercontent.com/5098581/187353254-8ae07fc7-3565-4e14-8189-93112c6900c3.png)

Another 5 minutes later, no change. I disabled "Noise Reduction" and the Gate effect is still closed, but something (artificial?) is now getting boosted by "Autogain":

![Screenshot_20220830_170406](https://user-images.githubusercontent.com/5098581/187353565-a91814a3-d934-4f52-a8ee-9959b5cee248.png)

![Screenshot_20220830_170622](https://user-images.githubusercontent.com/5098581/187353933-b4af2d71-6cce-405f-ae34-86207ab72d4d.png)

No change to _Output Gain_ until a louder sound is received through the Gate opening that boosted the value, no reduction:

![Screenshot_20220830_170815](https://user-images.githubusercontent.com/5098581/187354105-ed325a1f-ab14-4280-af6e-cf7d6f86729b.png)

The signal from disabling RNNoise is actually appearing due to "Limiter" + "Autogain", disabling either of those drops it back down to silent:

![Screenshot_20220830_171016](https://user-images.githubusercontent.com/5098581/187354513-fea76726-6edf-41c4-b952-794f7a9c108c.png)

Looks like I was wrong about the UI bar rendering differently when not overflowing, at least with the Gate blocking most noise currently, _Output Gain_ remains fixed at this value:

![Screenshot_20220830_171343](https://user-images.githubusercontent.com/5098581/187355054-b7701fc8-dfb9-47b6-8c9e-09dd764dbf8f.png)

Used both reset buttons, presumably _Output Gain_ does not change since both _Input_ and _Output_ are 99db?:

![Screenshot_20220830_171744](https://user-images.githubusercontent.com/5098581/187355424-4300f25b-9776-4822-93a7-b6e0176f175f.png)

</details>

Let me know if you'd like me to file a bug report for that (Reset not setting _Output Gain_ back to 0). I know there is one for the UI rendering issue with GTK4, but assumed that didn't affect the rendered value getting stuck.

**UPDATE:** I still think it's odd that _Output Gain_ would not lower over time, but I guess it's trying to maintain the low volume idle signal at the target volume? (_even when a Gate cuts that signal_)

In practice it's fine so long as a Gate is present to block out any idle noise.

### Limiter

> Yes. I usually have a limiter as the last filter and the autogain before it. This way the limiter can protect the soundcard if the autogain does something odd.

Awesome :+1: 

Just to confirm, the Limiter still provides this protection if setting a boost? (_I think I was using that as a workaround when I had "Autogain" disabled_).

Reading the in-app manual page, the limiter would boost the signal by `-40db` and is continues to apply at `0db`, so I think my ears are safe :smile: 

Alternatively without the boost I can lower that range, although I'm not sure how useful that is. I had read with OBS the meter (with green, yellow, red zones) that your voice should ideally be hitting the yellow zone, so removing the boost and using `-5db` or so seems to help avoid some spill over into red (above -10db I think).

I am not sure if EasyEffects provides a similar meter to visualize / monitor, but perhaps that's not necessary thanks to Autogain + Limiter.

### Compressor speaker output + Sidechain on mic input

> You can achieve the same thing if you put in your speakers pipeline a compressor with the sidechain input set to your microphone.
> This way when you speak the volume of whatever you are listening to will be automatically reduced.

Great tip, cheers :grinning:

**Reply by wwmm** [reply]
>I still think it's odd that Output Gain would not lower over time, but I guess it's trying to maintain the low volume idle signal at the target volume? 

Yes. I have to investigate this. The autogain should have stopped messing with the output gain once the momentary loudness was below `-70`. But for some reason it is still trying to make the input signal, that is literally just silence, to reach the target loudness. That is why the output gain is so high.

>Just to confirm, the Limiter still provides this protection if setting a boost?

It should still keep the output signal below `0 dB`. At least in some of the tests I did now that was what happened. But have in mind the limiter is a last resort for protection. If you actually push it to its limits you will have changes to the signal dynamic range that usually do not sound good even if the card is protected.

As I do not use the boost function there may be some details I am not aware of. Usually when I Want to make aggreesive changes to the dynamic range I use the `maximizer` plugin. It is more straightforward to use and is enough for my needs.

**Reply by wwmm** [reply]
>The autogain should have stopped messing with the output gain once the momentary loudness was below -70

I am doing some tests and this happens. But if the signal momentary loudness stays below `-70` the autogain will keep whatever it was the last gain value. That is why the gain levelbar stays forever in the same way. The autogain is waiting for the signal to get out of the `silence` range to mess with the gain again. This is fine. What I have to do is allowing the user to customize the `silence threshold` to values different than -70.

**Comment by polarathene** [comment, marked as answer]
I'd share this as a preset, but I don't see a convenient way to export via the UI, and the in-app manual entry doesn't mention a location.

I know it's somewhere, I'd just have to go look in this repo or on my system to find it :sweat_smile: 

**I've no idea how useful my settings below are for others to reference.** 

<details>
<summary>NOTE: Physical mic settings</summary>

The Deity VO-7U dynamic mic I use has the gain knob set to half-way which is barely audible at 20% volume on my headphones without Autogain (_which brings it to a volume that Youtube video speech is at for the same headphone volume, otherwise headphones need to go up to 100% volume to match the equivalent 20% level of Youtube_).

I could up the physical gain knob to compensate and rely less on Autogain boosting it, but that'd throw off all the settings below AFAIK, plus I live in a noisy environment, so I got a dynamic mic under the impression it'd reduce the more distant sound sources, and with the lower gain setting helping to filter that out better (_still audible with "Noise Reduction" and "Gate" disabled, "Autogain" makes that apparent unless I'm speaking_).

</details>

## Effects

<details>
<summary>RNNoise (Defaults)</summary>

### RNNoise (Defaults)

![Screenshot_20220901_180040](https://user-images.githubusercontent.com/5098581/187842546-fb6015f9-d44e-489f-985f-7a97ff7a78e8.png)

</details>

<details>
<summary>Gate</summary>

### Gate

I am not sure if I configured this correctly.

- It relies on RNNoise to decide if it's noise worth considering for the gate first (_usually if voice, but can also be the sound of coffee mug or similar being placed on my desk_).
- Threshold was adjusted to be responsive to voice (_I ended up looping some youtube video of someone speaking rather than myself, so it's not properly tuned to my voice yet, perhaps I could instead loop a recording somehow?_), while trying to block out other noises that might slip through RNNoise (_more likely if a stray sound relaxed it enough to not dampen out general background noise_)
- Attack is short, I found it worked nice with a longer attack to avoid stray sounds opening the gate, but that negatively impacted actual speaking, clipping the first parts of speech.
- Release fades off a bit more smoothly, it sounded odd if the gate was too aggressive at closing quickly.
- To avoid the issue with closing too quickly and allow for some freedom to move away from the mic. Autogain would normalize the volume level of (_I have a dynamic mic so falloff from distance would otherwise be much steeper_), I enabled the "Hysteris" settings which relax the Gates closing conditions so once open (assuming from speech) I didn't have to worry about the gate flipping on/off frequently (_more likely with RNNoise manipulating the signal?_), whilst low enough volume will close the gate (_especially if RNNoise also switches back to filtering out most of the signal again_).

**Side-chain was a bit tricky for me:**
- Mode: I liked _Low-Pass_, it'd better block out most noises, but also caused the gate to fluctuate more often or ignore some lower speech transitions like whispering / trailing off speech. Ended up sticking with RMS since it was more reliable at keeping the Gate open, even if it wasn't as favourable for filtering what opens the Gate.
- Lookahead: Is the only other setting I kept after experimenting. Honestly not sure if it's making much of a difference.
- Filters: I think the low/high pass filters here help filter the RNNoise audio that gets through to better trigger when actual speech comes through. I've read that bulk of speech information is in 100-300 with the higher frequencies being more complimentary, so I tried to lower the side-chain filter to listening to that range. It was less reliable below 1kHz, but this seems to work well :+1: 

![Screenshot_20220901_173521](https://user-images.githubusercontent.com/5098581/187839151-a04380b6-cbfb-4b60-ad9b-1057e6c822fd.png)

**Side-chain:**

![Screenshot_20220901_173544](https://user-images.githubusercontent.com/5098581/187839186-d8801e66-5172-4d42-bc03-ac34c387b3af.png)

**Side-chain filters:**

![Screenshot_20220901_173623](https://user-images.githubusercontent.com/5098581/187839279-b3d9243a-f7c5-481f-b4c5-c10502701ddc.png)

</details>

<details>
<summary>Autogain (Defaults)</summary>

### Autogain (Defaults)

Defaults seemed to work well / reliably after experimenting a fair bit.

![Screenshot_20220901_173446](https://user-images.githubusercontent.com/5098581/187839070-a0723dcc-ef86-44bb-b30d-7f688fcb383c.png)

</details>

<details>
<summary>Equalizer</summary>

### Equalizer

This is probably not great, I winged it (_again with the youtube video playback with several vocal ranges_). It sounded a little better for the various voices running through it, but not tuned to a specific one.

Eventually I'll try tune it to my own to see if it makes much of a difference, and keep the proper frequency bands in mind, instead of play-by-ear tweaking :sweat_smile: 

![Screenshot_20220901_173354](https://user-images.githubusercontent.com/5098581/187838956-bf90164a-a4d7-46ff-8eff-324d405cc338.png)

I wouldn't have minded adding some new bands, but that seems to not be good UX to adjust for presently. I used the same tweaking approach as with most effects.

</details>

<details>
<summary>Limiter</summary>

### Limiter

Mostly defaults, except for Mode, Oversampling and Dither. Also enabled "Auto Leveling".

Again, little idea what I'm doing. At best I looked at the in-app manual for explaining some settings and then trying them out to try hear what sounded better.
- Mode somewhat had an effect, easier to hear if doing something loud like light tapping on the mic itself (_hopefully safe to the equipment_).
- Oversampling I couldn't hear any notable difference.
- Dither might be why I couldn't. I know that 8-bit sounds worse than None, but 24-bit also sounded alright.

![Screenshot_20220901_173323](https://user-images.githubusercontent.com/5098581/187838897-9df057db-482f-4e61-a1cc-b27f4a649b37.png)

</details>

---

Hopefully it's a little helpful :+1:

**Comment by wucke13** [comment]
The profile files can be found under `~/.config/easyeffects/input/`. I copied the settings above more or less 1:1 except that I added mono to stereo upmix at the very end. Find attached the file: [microphone.json](https://github.com/user-attachments/files/17803250/microphone.json)
